### ISSUE TITLE
How to perform batch inserts to clickhouse using JDBC client?

### ISSUE THREAD

**Issue by kvbfullstack** [ClickHouse/clickhouse-java#1392]
I am trying to write into clickhouse DB using Clickhouse JDBC client (0.4.0)

`<dependency>`
`    <groupId>com.clickhouse</groupId>`
`    <artifactId>clickhouse-jdbc</artifactId>`
`    <version>0.4.0</version>`
`</dependency>`

[I am following this example to do batch insert](https://github.com/ClickHouse/ClickHouse/issues/33245#issuecomment-1003579177)

Basically,
`try (PreparedStatement ps = conn.prepareStatement(`
`    "insert into mytable select col1, col2 from input('col1 String, col2 DateTime64(3), col3 Int32')")) {`
`    ps.setString(1, "test");`
`    ps.setObject(2, LocalDateTime.now());`
`    ps.setInt(3, 123);`
`    ps.addBatch();`
`    ...`
`    ps.addBatch();`
`    ...`
`    ps.addBatch();`
`    ...`
`    ps.executeBatch();`
`}`

I am adding 10000 rows to the batch before executing `ps.executeBatch()`. 
Is it the right and correct way to do batch inserts to clickhouse ?

Is it similar as doing ? `insert into mytable values(...)(...)...(...)`

**Comment by zhicwu** [comment]
Hi @kvbfullstack, using input function in batch insert is something like the following. Behind the scene, there are chunked requests being sent to ClickHouse server for insertion.

```sql
insert into mytable format RowBinary
<serialized data in RowBinary format>
```

RE: batch size, it depends on partions, columns, and the volume. You may increase the batch size by multiply 10 and watch if server is comfortable to handle that.

**Comment by kvbfullstack** [comment]
Hi @zhicwu , thanks for the clarification.

Are you suggesting that the when we call `ps.executeBatch()`, it may result into one or more `insert into mytable format RowBinary` statements  depending on the other attributes you mentioned (partitions, columns, and the volume) ?

**Comment by zhicwu** [comment]
By calling `ps.executeBatch()`, it actually results in one `insert into mytable format RowBinary` statement, in one or more [chunks](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) in http request. `addBatch` on the other hand just sends the serialized parameters to a pipeline, which will be later send to server.

If you're seeking for a streaming approach, you may use `ClickHouseWriter` instead, as demonstrated below:

https://github.com/ClickHouse/clickhouse-java/blob/c25ffc62c886c0cdc1fcf8d8bf2234690c131dca/examples/jdbc/src/main/java/com/clickhouse/examples/jdbc/Basic.java#L118-L137

Alternatively, if the data was first generated in a file like JSON/CSV etc., you may use `insert into test_in_and_out from infile 'my.csv.zst' compression 'lz4' format TSV`, which is faster than above two approaches due to no serialization in Java - it's all taking care of by ClickHouse server.

Lastly, if you're working on something which will be used in a production environment, please consider to upgrade JDBC driver to v0.4.6 which contains the most recent bugs fixes compare to 0.4.0.

**Comment by kvbfullstack** [comment]
Thanks for the great explanation @zhicwu
